# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


In [54]:
import pandas as pd

import numpy as np

df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")

In [56]:
df.shape

(4008, 11)

In [58]:
df.isna().sum()

Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

In [60]:
df.head(2)

Customer          ST GENDER Education Customer Lifetime Value  Income  \
0  RB50392  Washington    NaN    Master                     NaN     0.0   
1  QZ44356     Arizona      F  Bachelor              697953.59%     0.0   

   Monthly Premium Auto Number of Open Complaints    Policy Type  \
0                1000.0                    1/0/00  Personal Auto   
1                  94.0                    1/0/00  Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935

In [62]:
df.tail(2)

Customer   ST GENDER Education Customer Lifetime Value  Income  \
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
4006                   NaN                       NaN         NaN   
4007                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
4006           NaN                 NaN  
4007           NaN                 NaN

In [64]:
df.sample(3)

Customer   ST GENDER Education Customer Lifetime Value  Income  \
3846      NaN  NaN    NaN       NaN                     NaN     NaN   
2472      NaN  NaN    NaN       NaN                     NaN     NaN   
2598      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
3846                   NaN                       NaN         NaN   
2472                   NaN                       NaN         NaN   
2598                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
3846           NaN                 NaN  
2472           NaN                 NaN  
2598           NaN                 NaN

# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [12]:
# df.columns # columns before any changes

In [66]:
df.columns = [col.lower().replace(" ", "_") for col in df.columns]

In [68]:
df.rename(columns= {'st': 'state'}, inplace=True)

In [70]:
df.columns

Index(['customer', 'state', 'gender', 'education', 'customer_lifetime_value',
       'income', 'monthly_premium_auto', 'number_of_open_complaints',
       'policy_type', 'vehicle_class', 'total_claim_amount'],
      dtype='object')

## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:
- *Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*
- *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*
- *In education, "Bachelors" could be replaced by "Bachelor"*
- *In Customer Lifetime Value, delete the `%` character*
- *In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [72]:
df.gender.unique() # categories before any changes

array([nan, 'F', 'M', 'Femal', 'Male', 'female'], dtype=object)

In [74]:
df.gender.replace("Femal", "F", inplace=True)
df.gender.replace("female", "F", inplace=True)
df.gender.replace("Male", "M", inplace=True)

df.gender.value_counts()

gender
F    502
M    452
Name: count, dtype: int64

In [76]:
df.state.unique() # categories before any changes

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', 'Cali',
       'AZ', 'WA', nan], dtype=object)

In [78]:
df.state.replace("Cali", "California", inplace=True)
df.state.replace("AZ", "Arizona", inplace=True)
df.state.replace("WA", "Washington", inplace=True)

# df['state'] = df['state'].map({'Cali': 'California', 'AZ': 'Arizona', 'WA': 'Washington'})
# if done like this: not mentioned categories are deleted

df.state.unique()

array(['Washington', 'Arizona', 'Nevada', 'California', 'Oregon', nan],
      dtype=object)

In [80]:
df.education.unique() # categories before any changes

array(['Master', 'Bachelor', 'High School or Below', 'College',
       'Bachelors', 'Doctor', nan], dtype=object)

In [82]:
df.education.replace("Bachelors", "Bachelor", inplace=True)

df.education.value_counts()

education
Bachelor                331
College                 313
High School or Below    296
Master                   94
Doctor                   37
Name: count, dtype: int64

In [84]:
df["customer_lifetime_value"] = df["customer_lifetime_value"].astype(str).str.replace("%", "") # convert to str to enable str formatting

df["customer_lifetime_value"] = pd.to_numeric(df["customer_lifetime_value"], errors='coerce') # change back to numeric

df.customer_lifetime_value.sample(5)

2667          NaN
3034          NaN
1017    501208.37
2535          NaN
1735          NaN
Name: customer_lifetime_value, dtype: float64

In [86]:
df.vehicle_class.unique() # categories before any changes

array(['Four-Door Car', 'Two-Door Car', 'SUV', 'Luxury SUV', 'Sports Car',
       'Luxury Car', nan], dtype=object)

In [88]:
df.vehicle_class.replace("Sports Car", "Luxury", inplace=True)
df.vehicle_class.replace("Luxury SUV", "Luxury", inplace=True)
df.vehicle_class.replace("Luxury Car", "Luxury", inplace=True)

df.vehicle_class.value_counts()

vehicle_class
Four-Door Car    576
Two-Door Car     205
SUV              199
Luxury            91
Name: count, dtype: int64

## Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [142]:
# dropping rows with only NaNs in them
# doing this now already because without this, I cannot do open_complaints (below)

df.dropna(how='all', inplace=True) 

In [144]:
df.isna().sum() # only two columns with NaNs left: "gender" and "customer_lifetime_value"

customer                       0
state                          0
gender                       117
education                      0
customer_lifetime_value        0
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64

In [146]:
df.shape

(1071, 11)

In [98]:
df.sample(2)

customer       state gender education  customer_lifetime_value   income  \
1041  QS75550  Washington      F  Bachelor                248004.59  93383.0   
568   VM13430      Oregon      F  Bachelor                860915.82  79090.0   

      monthly_premium_auto number_of_open_complaints     policy_type  \
1041                  62.0                    1/0/00  Corporate Auto   
568                  107.0                    1/0/00  Corporate Auto   

     vehicle_class  total_claim_amount  
1041  Two-Door Car          244.212286  
568            SUV          289.040734

In [100]:
df.dtypes

customer                      object
state                         object
gender                        object
education                     object
customer_lifetime_value      float64
income                       float64
monthly_premium_auto         float64
number_of_open_complaints     object
policy_type                   object
vehicle_class                 object
total_claim_amount           float64
dtype: object

"""
customer_lifetime_value was already dealt with before when erasing the '%', dtype was changed to float
"""

In [102]:
df.number_of_open_complaints.unique()

array(['1/0/00', '1/2/00', '1/1/00', '1/3/00', '1/5/00', '1/4/00'],
      dtype=object)

In [104]:
df["number_of_open_complaints"] = df["number_of_open_complaints"].astype(str).str.split("/")

df["number_of_open_complaints"] = df["number_of_open_complaints"].apply(lambda complaints: int(complaints[1]))

df["number_of_open_complaints"] = pd.to_numeric(df["number_of_open_complaints"], errors='coerce')

df.number_of_open_complaints

0       0
1       0
2       0
3       0
4       0
       ..
1066    1
1067    2
1068    0
1069    0
1070    0
Name: number_of_open_complaints, Length: 1071, dtype: int64

## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers.

rows with only NaNs in it were dropped already with df.dropna(how='all', inplace=True), see beginning of Exercise 3

In [148]:
mean_clv = round(df.customer_lifetime_value.mean(), 2)
mean_clv

793114.58

In [150]:
median_clv = df.customer_lifetime_value.median()
median_clv

588174.235

In [152]:
# after checking median and mean, I decide to use the median to replace the 3 NaNs in customer_lifetime_value
# especially because the mean seems to be influenced a lot by outliers (unusually much higher than median)

df["customer_lifetime_value"] = df["customer_lifetime_value"].fillna(median_clv)

df.customer_lifetime_value.isna().sum()

0

In [154]:
df.gender.value_counts()

gender
F    502
M    452
Name: count, dtype: int64

In [156]:
round(df.gender.value_counts() / df.shape[0] * 100, 2)

gender
F    46.87
M    42.20
Name: count, dtype: float64

In [158]:
gender_prob = df["gender"].value_counts(normalize=True)

In [160]:
gender_prob

gender
F    0.526205
M    0.473795
Name: proportion, dtype: float64

"""
One option to replace NaNs in gender would be ffill or bfill but a quick search in head() (see below) shows that many NaNs seems to be in groups (only 117 all together, so those roughly more than 30 in the head() are relevant). If filled with bfill or ffill, these would all take on the value of the one before/after and thus is not a random distribution.

Another (even worse) option would be to replace the NaNs with the mode, this would falsify the analysis because all missing values would be counted as female which is highly unlikely.

A distribution by % would be best, see above: roughly 47% female and 42% male. So we would need to calc the % of f/m w/o the NaNs (see above, done with variable "gender_prob") and then fill NaNs accordingly. But I don't know how to do that and I could find only ways which I don't fully understand. (Like this: df["gender"] = df['gender'].fillna(pd.Series(np.random.choice([],  p=8], size=len(df))))

In [162]:
df.gender.head(50)

0     NaN
1       F
2       F
3       M
4       M
5       F
6       F
7       M
8       M
9       F
10      M
11    NaN
12    NaN
13    NaN
14    NaN
15    NaN
16    NaN
17    NaN
18    NaN
19    NaN
20    NaN
21      M
22    NaN
23    NaN
24    NaN
25    NaN
26    NaN
27      M
28    NaN
29    NaN
30    NaN
31    NaN
32    NaN
33      F
34    NaN
35    NaN
36    NaN
37    NaN
38    NaN
39    NaN
40    NaN
41    NaN
42    NaN
43    NaN
44      M
45    NaN
46    NaN
47    NaN
48    NaN
49    NaN
Name: gender, dtype: object

## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [164]:
df.duplicated().sum()

0

""" No duplicates here. If there were any, I would most likely proceed as follows:
1. keeping the first occurence of each duplicate, drop every repetition (duplicate)
2. I would not drop any data based on a subset of columns or based on a specific column

In [166]:
df.reset_index(drop=True, inplace=True) # index was fine (as no duplicates were dropped) but this is the method to reset
df.tail(5)

customer       state gender             education  \
1066  TM65736      Oregon      M                Master   
1067  VJ51327  California      F  High School or Below   
1068  GS98873     Arizona      F              Bachelor   
1069  CW49887  California      F                Master   
1070  MY31220  California      F               College   

      customer_lifetime_value   income  monthly_premium_auto  \
1066                305955.03  38644.0                  78.0   
1067               2031499.76  63209.0                 102.0   
1068                323912.47  16061.0                  88.0   
1069                462680.11  79487.0                 114.0   
1070                899704.02  54230.0                 112.0   

      number_of_open_complaints    policy_type  vehicle_class  \
1066                          1  Personal Auto  Four-Door Car   
1067                          2  Personal Auto            SUV   
1068                          0  Personal Auto  Four-Door Car   
1069                          0   Special Auto            SUV   
1070                          0  Personal Auto   Two-Door Car   

      total_claim_amount  
1066          361.455219  
1067          207.320041  
1068          633.600000  
1069          547.200000  
1070          537.600000

In [168]:
df.to_csv("df.csv", index=False) # saving data as csv, no new index

Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the isnull() function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

Drop the rows or columns with null values
Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
Fill the null values with the previous or next value in the column
Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)
Implement your chosen strategy to handle the null values. You can use the fillna() function in pandas to fill null values or dropna() function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the isnull() function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers.

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [ ]:
# Your code here